File: app/src/testUltimate/java/com/clockmods/ultimate/clock/UltimateClockPreferencesTest.java
package com.clockmods.ultimate.clock;

import android.content.SharedPreferences;

import com.clockmods.sdk.clock.ClockState;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class UltimateClockPreferencesTest {
    private MemorySharedPreferences stored;
    private UltimateClockPreferences preferences;

    @Before
    public void setUp() {
        stored = new MemorySharedPreferences();
        preferences = new UltimateClockPreferences(stored);
    }

    @Test
    public void defaultsMatchSettingsContract() {
        Assert.assertEquals("clockmods_ultimate_style",
                UltimateClockPreferences.PREFERENCES_NAME);
        Assert.assertEquals("style_id", UltimateClockPreferences.KEY_STYLE_ID);
        Assert.assertEquals("second_motion", UltimateClockPreferences.KEY_SECOND_MOTION);
        Assert.assertEquals("follow_reduced_motion",
                UltimateClockPreferences.KEY_FOLLOW_REDUCED_MOTION);
        Assert.assertEquals("background_mode", UltimateClockPreferences.KEY_BACKGROUND_MODE);
        Assert.assertEquals(UltimateClockStyles.STYLE_GLASS_ATELIER,
                preferences.getStyleId());
        Assert.assertEquals(ClockState.SecondHandMotion.SWEEP,
                preferences.getSecondHandMotion());
        Assert.assertTrue(preferences.isFollowSystemReducedMotion());
        Assert.assertEquals(UltimateClockPreferences.BACKGROUND_MODE_THEME,
                preferences.getBackgroundMode());
    }

    @Test
    public void settingsMotionStringsMapToSdkValues() {
        assertStoredMotion("smooth", ClockState.SecondHandMotion.SWEEP);
        assertStoredMotion("tick", ClockState.SecondHandMotion.TICK);
        assertStoredMotion("off", ClockState.SecondHandMotion.OFF);
        assertStoredMotion("unknown", UltimateClockPreferences.DEFAULT_SECOND_HAND_MOTION);
    }

    @Test
    public void sdkMotionValuesPersistAsSettingsStrings() {
        preferences.setSecondHandMotion(ClockState.SecondHandMotion.SWEEP);
        Assert.assertEquals("smooth", stored.getString(
                UltimateClockPreferences.KEY_SECOND_MOTION, null));
        preferences.setSecondHandMotion(ClockState.SecondHandMotion.TICK);
        Assert.assertEquals("tick", stored.getString(
                UltimateClockPreferences.KEY_SECOND_MOTION, null));
        preferences.setSecondHandMotion(ClockState.SecondHandMotion.OFF);
        Assert.assertEquals("off", stored.getString(
                UltimateClockPreferences.KEY_SECOND_MOTION, null));
    }

    @Test
    public void dottedStyleIdsRoundTripAndBlankValuesUseDefault() {
        preferences.setStyleId(UltimateClockStyles.STYLE_ORBIT_NEON);
        Assert.assertEquals("orbit.neon", stored.getString(
                UltimateClockPreferences.KEY_STYLE_ID, null));
        Assert.assertEquals("orbit.neon", preferences.getStyleId());

        preferences.setStyleId("   ");
        Assert.assertEquals(UltimateClockPreferences.DEFAULT_STYLE_ID,
                preferences.getStyleId());
    }

    @Test
    public void styleIdNormalizationRejectsEmptyAndOversizedValues() {
        Assert.assertEquals(UltimateClockPreferences.DEFAULT_STYLE_ID,
                UltimateClockPreferences.normalizeStyleId(null));
        Assert.assertEquals(UltimateClockPreferences.DEFAULT_STYLE_ID,
                UltimateClockPreferences.normalizeStyleId("  "));
        Assert.assertEquals("partner.meridian",
                UltimateClockPreferences.normalizeStyleId("  partner.meridian  "));
        StringBuilder oversized = new StringBuilder();
        for (int i = 0; i < 121; i++) oversized.append('a');
        Assert.assertEquals(UltimateClockPreferences.DEFAULT_STYLE_ID,
                UltimateClockPreferences.normalizeStyleId(oversized.toString()));
    }

    @Test
    public void restoreDefaultsWritesAllFourSettings() {
        preferences.setStyleId(UltimateClockStyles.STYLE_DIGITAL_GRID);
        preferences.setSecondHandMotion(ClockState.SecondHandMotion.OFF);
        preferences.setFollowSystemReducedMotion(false);
        preferences.setBackgroundMode(UltimateClockPreferences.BACKGROUND_MODE_IMAGE);

        preferences.restoreDefaults();

        Assert.assertEquals(UltimateClockPreferences.DEFAULT_STYLE_ID,
                preferences.getStyleId());
        Assert.assertEquals(UltimateClockPreferences.DEFAULT_SECOND_HAND_MOTION,
                preferences.getSecondHandMotion());
        Assert.assertEquals(UltimateClockPreferences.DEFAULT_FOLLOW_SYSTEM_REDUCED_MOTION,
                preferences.isFollowSystemReducedMotion());
        Assert.assertEquals(UltimateClockPreferences.DEFAULT_BACKGROUND_MODE,
                preferences.getBackgroundMode());
    }

    @Test
    public void backgroundModeAcceptsOnlyThreeStableValues() {
        preferences.setBackgroundMode(UltimateClockPreferences.BACKGROUND_MODE_COLOR);
        Assert.assertEquals(UltimateClockPreferences.BACKGROUND_MODE_COLOR,
                preferences.getBackgroundMode());
        preferences.setBackgroundMode(UltimateClockPreferences.BACKGROUND_MODE_IMAGE);
        Assert.assertEquals(UltimateClockPreferences.BACKGROUND_MODE_IMAGE,
                preferences.getBackgroundMode());
        preferences.setBackgroundMode("unknown");
        Assert.assertEquals(UltimateClockPreferences.BACKGROUND_MODE_THEME,
                preferences.getBackgroundMode());
    }

    private void assertStoredMotion(String storedValue, ClockState.SecondHandMotion expected) {
        stored.edit().putString(UltimateClockPreferences.KEY_SECOND_MOTION, storedValue).commit();
        Assert.assertEquals(expected, preferences.getSecondHandMotion());
    }

    private static final class MemorySharedPreferences implements SharedPreferences {
        private final Map<String, Object> values = new HashMap<String, Object>();

        @Override public Map<String, ?> getAll() {
            return Collections.unmodifiableMap(new HashMap<String, Object>(values));
        }

        @Override public String getString(String key, String defaultValue) {
            Object value = values.get(key);
            return value instanceof String ? (String) value : defaultValue;
        }

        @SuppressWarnings("unchecked")
        @Override public Set<String> getStringSet(String key, Set<String> defaultValues) {
            Object value = values.get(key);
            return value instanceof Set
                    ? new HashSet<String>((Set<String>) value) : defaultValues;
        }

        @Override public int getInt(String key, int defaultValue) {
            Object value = values.get(key);
            return value instanceof Integer ? (Integer) value : defaultValue;
        }

        @Override public long getLong(String key, long defaultValue) {
            Object value = values.get(key);
            return value instanceof Long ? (Long) value : defaultValue;
        }

        @Override public float getFloat(String key, float defaultValue) {
            Object value = values.get(key);
            return value instanceof Float ? (Float) value : defaultValue;
        }

        @Override public boolean getBoolean(String key, boolean defaultValue) {
            Object value = values.get(key);
            return value instanceof Boolean ? (Boolean) value : defaultValue;
        }

        @Override public boolean contains(String key) {
            return values.containsKey(key);
        }

        @Override public Editor edit() {
            return new MemoryEditor();
        }

        @Override public void registerOnSharedPreferenceChangeListener(
                OnSharedPreferenceChangeListener listener) { }

        @Override public void unregisterOnSharedPreferenceChangeListener(
                OnSharedPreferenceChangeListener listener) { }

        private final class MemoryEditor implements Editor {
            private final Map<String, Object> updates = new HashMap<String, Object>();
            private final Set<String> removals = new HashSet<String>();
            private boolean clear;

            @Override public Editor putString(String key, String value) {
                updates.put(key, value);
                return this;
            }

            @Override public Editor putStringSet(String key, Set<String> value) {
                updates.put(key, value == null ? null : new HashSet<String>(value));
                return this;
            }

            @Override public Editor putInt(String key, int value) {
                updates.put(key, value);
                return this;
            }

            @Override public Editor putLong(String key, long value) {
                updates.put(key, value);
                return this;
            }

            @Override public Editor putFloat(String key, float value) {
                updates.put(key, value);
                return this;
            }

            @Override public Editor putBoolean(String key, boolean value) {
                updates.put(key, value);
                return this;
            }

            @Override public Editor remove(String key) {
                removals.add(key);
                return this;
            }

            @Override public Editor clear() {
                clear = true;
                return this;
            }

            @Override public boolean commit() {
                apply();
                return true;
            }

            @Override public void apply() {
                if (clear) values.clear();
                for (String key : removals) values.remove(key);
                for (Map.Entry<String, Object> entry : updates.entrySet()) {
                    if (entry.getValue() == null) values.remove(entry.getKey());
                    else values.put(entry.getKey(), entry.getValue());
                }
            }
        }
    }
}

File: app/src/testUltimate/java/com/clockmods/ultimate/clock/UltimateClockPreferencesTest.kt
package com.clockmods.ultimate.clock

import android.content.SharedPreferences
import com.clockmods.sdk.clock.ClockState
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

class UltimateClockPreferencesTest {
    private lateinit var stored: MemorySharedPreferences
    private lateinit var preferences: UltimateClockPreferences

    @Before
    fun setUp() {
        stored = MemorySharedPreferences()
        preferences = UltimateClockPreferences(stored)
    }

    @Test
    fun defaultsMatchSettingsContract() {
        assertEquals("clockmods_ultimate_style", UltimateClockPreferences.PREFERENCES_NAME)
        assertEquals("style_id", UltimateClockPreferences.KEY_STYLE_ID)
        assertEquals("second_motion", UltimateClockPreferences.KEY_SECOND_MOTION)
        assertEquals("background_mode", UltimateClockPreferences.KEY_BACKGROUND_MODE)
        assertEquals(UltimateClockStyles.STYLE_GLASS_ATELIER, preferences.getStyleId())
        assertEquals(ClockState.SecondHandMotion.SWEEP, preferences.getSecondHandMotion())
        assertEquals(UltimateClockPreferences.BACKGROUND_MODE_THEME, preferences.getBackgroundMode())
    }

    @Test
    fun settingsMotionStringsMapToSdkValues() {
        assertStoredMotion("smooth", ClockState.SecondHandMotion.SWEEP)
        assertStoredMotion("tick", ClockState.SecondHandMotion.TICK)
        assertStoredMotion("off", ClockState.SecondHandMotion.OFF)
        assertStoredMotion("unknown", UltimateClockPreferences.DEFAULT_SECOND_HAND_MOTION)
    }

    @Test
    fun sdkMotionValuesPersistAsSettingsStrings() {
        preferences.setSecondHandMotion(ClockState.SecondHandMotion.SWEEP)
        assertEquals("smooth", stored.getString(UltimateClockPreferences.KEY_SECOND_MOTION, null))
        preferences.setSecondHandMotion(ClockState.SecondHandMotion.TICK)
        assertEquals("tick", stored.getString(UltimateClockPreferences.KEY_SECOND_MOTION, null))
        preferences.setSecondHandMotion(ClockState.SecondHandMotion.OFF)
        assertEquals("off", stored.getString(UltimateClockPreferences.KEY_SECOND_MOTION, null))
    }

    @Test
    fun dottedStyleIdsRoundTripAndBlankValuesUseDefault() {
        preferences.setStyleId(UltimateClockStyles.STYLE_ORBIT_NEON)
        assertEquals("orbit.neon", stored.getString(UltimateClockPreferences.KEY_STYLE_ID, null))
        assertEquals("orbit.neon", preferences.getStyleId())
        preferences.setStyleId("   ")
        assertEquals(UltimateClockPreferences.DEFAULT_STYLE_ID, preferences.getStyleId())
    }

    @Test
    fun styleIdNormalizationRejectsEmptyAndOversizedValues() {
        assertEquals(UltimateClockPreferences.DEFAULT_STYLE_ID, UltimateClockPreferences.normalizeStyleId(null))
        assertEquals(UltimateClockPreferences.DEFAULT_STYLE_ID, UltimateClockPreferences.normalizeStyleId("  "))
        assertEquals("partner.meridian", UltimateClockPreferences.normalizeStyleId("  partner.meridian  "))
        assertEquals(UltimateClockPreferences.DEFAULT_STYLE_ID, UltimateClockPreferences.normalizeStyleId("a".repeat(121)))
    }

    @Test
    fun restoreDefaultsWritesAllSettings() {
        preferences.setStyleId(UltimateClockStyles.STYLE_DIGITAL_GRID)
        preferences.setSecondHandMotion(ClockState.SecondHandMotion.OFF)
        preferences.setBackgroundMode(UltimateClockPreferences.BACKGROUND_MODE_IMAGE)
        preferences.restoreDefaults()
        assertEquals(UltimateClockPreferences.DEFAULT_STYLE_ID, preferences.getStyleId())
        assertEquals(UltimateClockPreferences.DEFAULT_SECOND_HAND_MOTION, preferences.getSecondHandMotion())
        assertEquals(UltimateClockPreferences.DEFAULT_BACKGROUND_MODE, preferences.getBackgroundMode())
    }

    @Test
    fun backgroundModeAcceptsOnlyThreeStableValues() {
        preferences.setBackgroundMode(UltimateClockPreferences.BACKGROUND_MODE_COLOR)
        assertEquals(UltimateClockPreferences.BACKGROUND_MODE_COLOR, preferences.getBackgroundMode())
        preferences.setBackgroundMode(UltimateClockPreferences.BACKGROUND_MODE_IMAGE)
        assertEquals(UltimateClockPreferences.BACKGROUND_MODE_IMAGE, preferences.getBackgroundMode())
        preferences.setBackgroundMode("unknown")
        assertEquals(UltimateClockPreferences.BACKGROUND_MODE_THEME, preferences.getBackgroundMode())
    }

    @Test
    fun palettesPersistIndependentlyAndResetWithDefaults() {
        val orbit = UltimateClockStyles.STYLE_ORBIT
        val dual = UltimateClockStyles.STYLE_DUAL_BLOCKS
        preferences.setPalette(
            orbit,
            ClockPalette(0xFFFFFFFF.toInt(), 0xFF123456.toInt(), 0xFF000000.toInt())
                .withGaussianBlur(true).withBlurStrength(75).withBlurBrightness(68),
        )
        preferences.setPalette(dual, ClockPalette(0xFF345678.toInt(), 0xFFEEEEEE.toInt(), 0xFFFF0000.toInt()))
        val reopened = UltimateClockPreferences(stored)
        assertTrue(reopened.getPalette(orbit).gaussianBlur)
        assertEquals(75, reopened.getPalette(orbit).blurStrength)
        assertEquals(68, reopened.getPalette(orbit).blurBrightness)
        assertEquals(ClockPalette.DEFAULT.blurStrength, reopened.getPalette(dual).blurStrength)
        assertEquals(ClockPalette.DEFAULT.blurBrightness, reopened.getPalette(dual).blurBrightness)
        assertFalse(reopened.getPalette(dual).gaussianBlur)
        reopened.setBackgroundMode(UltimateClockPreferences.BACKGROUND_MODE_COLOR)
        assertTrue(reopened.getPalette(orbit).gaussianBlur)
        assertEquals(0xFFFFFFFF.toInt(), reopened.getPalette(orbit).background)
        assertEquals(0xFF000000.toInt(), reopened.getPalette(orbit).accent)
        assertEquals(0xFFEEEEEE.toInt(), reopened.getPalette(dual).panel)
        assertEquals(ClockPalette.DEFAULT.background, reopened.getPalette(UltimateClockStyles.STYLE_BLEND).background)
        reopened.restoreDefaults()
        assertFalse(reopened.getPalette(orbit).gaussianBlur)
        assertEquals(ClockPalette.DEFAULT.blurStrength, reopened.getPalette(orbit).blurStrength)
        assertEquals(ClockPalette.DEFAULT.blurBrightness, reopened.getPalette(orbit).blurBrightness)
        assertEquals(ClockPalette.DEFAULT.background, reopened.getPalette(orbit).background)
        assertEquals(ClockPalette.DEFAULT.panel, reopened.getPalette(dual).panel)
    }

    private fun assertStoredMotion(storedValue: String, expected: ClockState.SecondHandMotion) {
        stored.edit().putString(UltimateClockPreferences.KEY_SECOND_MOTION, storedValue).commit()
        assertEquals(expected, preferences.getSecondHandMotion())
    }

    private class MemorySharedPreferences : SharedPreferences {
        private val values = HashMap<String, Any?>()

        override fun getAll(): Map<String, *> = HashMap(values)
        override fun getString(key: String?, defaultValue: String?): String? = values[key] as? String ?: defaultValue
        @Suppress("UNCHECKED_CAST")
        override fun getStringSet(key: String?, defaultValues: Set<String>?): Set<String>? =
            (values[key] as? Set<String>)?.toHashSet() ?: defaultValues
        override fun getInt(key: String?, defaultValue: Int): Int = values[key] as? Int ?: defaultValue
        override fun getLong(key: String?, defaultValue: Long): Long = values[key] as? Long ?: defaultValue
        override fun getFloat(key: String?, defaultValue: Float): Float = values[key] as? Float ?: defaultValue
        override fun getBoolean(key: String?, defaultValue: Boolean): Boolean = values[key] as? Boolean ?: defaultValue
        override fun contains(key: String?): Boolean = values.containsKey(key)
        override fun edit(): SharedPreferences.Editor = MemoryEditor()
        override fun registerOnSharedPreferenceChangeListener(listener: SharedPreferences.OnSharedPreferenceChangeListener?) = Unit
        override fun unregisterOnSharedPreferenceChangeListener(listener: SharedPreferences.OnSharedPreferenceChangeListener?) = Unit

        private inner class MemoryEditor : SharedPreferences.Editor {
            private val updates = HashMap<String, Any?>()
            private val removals = HashSet<String>()
            private var clear = false

            override fun putString(key: String?, value: String?): SharedPreferences.Editor = put(key, value)
            override fun putStringSet(key: String?, value: Set<String>?): SharedPreferences.Editor = put(key, value?.toHashSet())
            override fun putInt(key: String?, value: Int): SharedPreferences.Editor = put(key, value)
            override fun putLong(key: String?, value: Long): SharedPreferences.Editor = put(key, value)
            override fun putFloat(key: String?, value: Float): SharedPreferences.Editor = put(key, value)
            override fun putBoolean(key: String?, value: Boolean): SharedPreferences.Editor = put(key, value)
            private fun put(key: String?, value: Any?): SharedPreferences.Editor {
                if (key != null) updates[key] = value
                return this
            }
            override fun remove(key: String?): SharedPreferences.Editor {
                if (key != null) removals += key
                return this
            }
            override fun clear(): SharedPreferences.Editor { clear = true; return this }
            override fun commit(): Boolean { apply(); return true }
            override fun apply() {
                if (clear) values.clear()
                removals.forEach(values::remove)
                updates.forEach { (key, value) -> if (value == null) values.remove(key) else values[key] = value }
            }
        }
    }
}
